Add "download all" buttons (including size of dataset) to dataset page

pdurbin opened the pull request:
**What this PR does / why we need it**:

This pull request adds "download all" buttons at the dataset level and shows the size of the dataset.

**Which issue(s) this PR closes**:

- Closes #6118  Make download all files feature more prominent
- Closes #6400 Display the total size of a dataset on the dataset page

**Special notes for your reviewer**:

(this should probably be removed, since the issue has been resolved? - At least I'm going to mark it as resolved. -- L.A.)
Riddle me this... why do I get a different "size of dataset" number that what I see for the file? Here's a screenshot:

![Screen Shot 2020-07-01 at 3 47 34 PM](https://user-images.githubusercontent.com/21006/86285243-3f5f3b80-bbb2-11ea-9535-f14bf247bb13.png)

Also, there has been discussion of adding a "download all files" API but @djbrooke and I have decided to split this off into a separate chunk, represented by #4529.

**Suggestions on how to test this**:

- Test both tabular and non tabular files.
- Test datasets with and without terms of use.
- Test datasets with and without a guestbook.
- Test restricted files.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

I don't see any mockups linked from #6118 but I believe they are our there somewhere. In addition to the screenshot above, here's one for a dataset with no tabular files:

![Screen Shot 2020-07-01 at 3 54 03 PM](https://user-images.githubusercontent.com/21006/86285707-1be8c080-bbb3-11ea-9f6b-efe7c413594f.png)

**Is there a release notes update needed for this change?**:

I think there's already a release note about general dataset redesign improvements? I can add another note if we want.

**Additional documentation**:

I didn't touch the User Guide because my understanding is that we are going to update it all at once after the UI changes.

mheppler commented on src/main/webapp/dataset.xhtml:
```diff
@@ -143,17 +139,29 @@
-                                                            <span class="ui-commandlink ui-widget ui-state-disabled btn-download">Download (Placeholder)</span>
+                                                        <!-- NORMAL DOWNLOAD BUTTON (NO TABULAR FILES) -->
+                                                        <li jsf:rendered="#{!DatasetPage.isHasTabular()}">
+                                                            <p:commandLink update="@form" actionListener="#{DatasetPage.validateAllFilesForDownloadArchival()}">
```
Links are missing the btn-download style class included in the placeholder, and required by the analytics code.

pdurbin replied:
Fixed in 054f249a4.

mheppler commented on src/main/webapp/dataset.xhtml:
```diff
@@ -143,17 +139,29 @@
+                                                        </li>
+                                                        <!-- DOWNLOAD ARCHIVAL FILES (TABULAR FILES PRESENT) -->
+                                                        <li jsf:rendered="#{DatasetPage.isHasTabular()}">
+                                                            <p:commandLink update="@form" actionListener="#{DatasetPage.validateAllFilesForDownloadArchival()}">
```
"ZIP" file type labels should be moved to the bundle.

mheppler replied:
There is already a `application/zip=ZIP Archive` entry in the MimeTypeDisplay.properties, not sure if it would be possible to pull from that file, or if we need a new entry in Bundle.properties.

pdurbin replied:
Fixed in 140ba41e4.

mheppler commented on src/main/java/edu/harvard/iq/dataverse/DatasetPage.java:
```diff
@@ -2886,8 +2889,35 @@ public void setSelectedDownloadableFiles(List<FileMetadata> selectedDownloadable
+    }
+
+    public void validateAllFilesForDownloadArchival() {
+        selectAllFiles();
```
I assume this `selectAllFiles()` is what's responsible for the odd UI behavior... when you select "Download" from the Access Dataset dropdown, you get the ZIP download, but now all the files are selected in the files table... that shouldn't be necessary... there has to be a way to get the ZIP without the unexpected UI behavior.

pdurbin replied:
I tried adding `clearSelection()` to the end like this...

```
     public void validateAllFilesForDownloadArchival() {
         selectAllFiles();
         boolean guestbookRequired = isDownloadPopupRequired();
         boolean downloadOriginal = false;
         validateFilesForDownload(guestbookRequired, downloadOriginal);
+        clearSelection();
     }
```

... but now it thinks no files are selected:

![Screen Shot 2020-07-02 at 1 08 07 PM](https://user-images.githubusercontent.com/21006/86389919-19df3a00-bc65-11ea-8a3a-2355a96d4b31.png)

So yes, I agree a fix would be nice but I'm not sure what that fix is.

scolapasta replied:
Wouldn't the fix be to use a different method than what the one that cares about selected files? i.e

change current method to get selected files and get a list, path that to new private method that takes core of current method but uses this new list.

Create a new method that generates a list and calls new that new private method.

You'd of course have to make sure all the checks for permissions are happening correctly, but I'd expect that to already be in the backend side (i.e. we already shouldn't be relying in the front end to only pass in the permissible values)

pdurbin replied:
@scolapasta @mheppler please see 427c88356. This pull request is now much bigger but it no longer uses `selectedFiles` which is what the checkboxes are connected to.

mheppler commented on src/main/java/edu/harvard/iq/dataverse/DatasetPage.java:
```diff
@@ -2886,8 +2889,35 @@ public void setSelectedDownloadableFiles(List<FileMetadata> selectedDownloadable
+        GetDatasetStorageSizeCommand cmd = new GetDatasetStorageSizeCommand(dvRequestService.getDataverseRequest(), dataset, false, GetDatasetStorageSizeCommand.Mode.DOWNLOAD, workingVersion);
+        try {
+            long bytes = commandEngine.submit(cmd);
+            return FileSizeChecker.bytesToHumanReadable(bytes);
```
Bug in the download size displayed. My locally tested datasets are saying the download will be "300 B", but the resulting ZIP is 11 KB, or 11,000 B.

pdurbin replied:
Yes, this what I noted when saying "riddle me this" in the description of this pull request (with a screenshot), that there's a mismatch in size. GetDatasetStorageSizeCommand needs discussion in tech hours or #dv-tech (heads up @scolapasta ) or a smaller group. I haven't looked into the implementation but the thing I was wondering about is how we estimate the size of a zip. Text files compress really well. Video files don't. I doubt the number will ever be 100% accurate. But yes, it seems pretty far off right now.

scolapasta replied:
We can discuss next tech hours, sure. But isn't the inaccuracy in the wrong direction?

i.e. the ip should be smaller than the size from GetDatasetStorageSizeCommand? I could imagine a scenario where it would be large (due to zip headers), but in that case, I would expect it to be close in size.

But maybe this is "close" in this scenario.

pdurbin replied:
As discussed in tech hours, the problem seems to stem from double counting tabular files (counting both the size of the original file and the archival version). I just pushed a fix in 19577769b.

mheppler commented:
Kicked the tires on the analytics a lil more. Built the branch, added style classes for `btn-download`, and confirmed that worked. Also realized that the explore like was missing the `btn-explore` style class, so the latest from develop needs to be merged as that style class was added in the recently merged **6938 analytics new buttons #7008** PR.

One additional detail that will need to be tweaked in the analytics code is to resolve how the new Access Dataset > Download clicks are recorded in the Google Analytics reports. The way the code works, it pulls the link text to report what button was clicked, so now that the file format and size was added to the new download link text, it makes into the report, which is going to mean you can not group those events by category (see screenshot).

<img width="815" alt="Screen Shot 2020-07-01 at 11 49 27 PM" src="https://user-images.githubusercontent.com/687227/86314381-86baeb80-bbf5-11ea-8646-9848b2afe2ba.png">

pdurbin commented:
I'm ready for more code review. I addresses the checkbox issue and am now differentiating between sizes for original files vs archival. Here's a screenshot:

<img width="342" alt="Screen Shot 2020-07-07 at 5 18 29 PM" src="https://user-images.githubusercontent.com/21006/86844420-07566d80-c076-11ea-812c-928da8e89a6e.png">

djbrooke commented:
Two more changes we want to make here, it was added to another issue by mistake. We believe the change to the publish button is already done.

<img width="609" alt="86605833-3507bd80-bf75-11ea-935b-bd58d2e3e382" src="https://user-images.githubusercontent.com/7773392/87324225-cdfe8180-c4fd-11ea-9890-c915c2e0e87f.png">

kcondon commented:
@pdurbin Any resolution to the file size download all mismatch? I'm seeing the zip download is larger than the individual files.

pdurbin commented:
@kcondon I thought the size estimate was better as of 1957776 (before that, tabular files were being counted twice). Can you please post a screenshot of the problem you're seeing?

kcondon commented:
So there are at least two issues:
1. the download size for both tab and original are off but the tab is off by a lot.
2. the actual size of the zip file is considerably smaller than what's presented in the download, so I am guessing it refers to the uncompressed size?

kcondon commented:
![image](https://user-images.githubusercontent.com/4983816/87351604-a91f0400-c527-11ea-9cec-8950724d0782.png)

kcondon commented:
This is a screen shot using FF, saving the zip to windows, downloading the original format:
![image](https://user-images.githubusercontent.com/4983816/87351755-e5526480-c527-11ea-9ccb-b6628a3d4be2.png)

landreev commented:
> So there are at least two issues:
> 
>     1. the download size for both tab and original are off but the tab is off by a lot.
> 
>     2. the actual size of the zip file is considerably smaller than what's presented in the download, so I am guessing it refers to the uncompressed size?
> 
> 
> Otherwise download all appears to work

Yes on 2. - that's not really an issue; the size the user is seeing in the download dialog is the compressed size of the zip; we want to report the sizes of the actual files. 

As for 1., looking at the sizes in the pull down menu in the example above: 
the size for the "original" format appears to be correct (62.3KB); but for the "tabular" it does appear to be off - actually, it looks like it's off by exactly the size of the original; i.e., it keeps counting both format sizes, instead of counting just the tabular size.

landreev commented:
After some testing and discussing it with Kevin - just another thought: 
maybe we want to add the word "uncompressed" there, to avoid confusion.
As in "(62.3KB Uncompressed)" instead of just "(62.3KB)"?

just an idea

mheppler commented:
> maybe we want to add the word "uncompressed" there, to avoid confusion "(62.3KB Uncompressed)"

Thank you for the review, @landreev. I thinking something along the same lines. Will review some options with @TaniaSchlatter and put suggested revisions here in this pull request.

pdurbin commented:
In 7e79f1b I addressed the problem tabular download options showing for versions of datasets that no longer have tabular files (because they were deleted and then the dataset was republished).

In 115353c I took another swing at getting GetDatasetStorageSizeCommand to behave properly and I thought I had it working but while testing I found further problems and weirdness so I ultimately decided to write a small command that does only the job I want. It seems to work fine so I switched over to it. See c13a5cc. I'm not doing any error checking within that small method (getDownloadSize in DatasetUtil) but I'm happy to add some if we're worried about nulls in the database for the size of files.

I'm moving this to code review.

@djbrooke and @TaniaSchlatter I noticed you're still assigned. I deployed the code here if you'd like to check it out but there are no UI changes: http://ec2-54-160-66-157.compute-1.amazonaws.com

mheppler commented:
Told Phil that I trusted his work, but I still went and confirmed the download sizes that I had poked at early. Looking much, much better. Didn't test anything larger than the 2.1 MB download here, but it be worth testing if a 210 MB download is off by 20 MB and if that is worth trying to fix.

<img width="1165" alt="Screen Shot 2020-07-16 at 5 00 27 PM" src="https://user-images.githubusercontent.com/687227/87722319-df59bf00-c785-11ea-997c-658e7992ce10.png">

landreev commented:
@mheppler @pdurbin The numbers above look right to me. What looks like a discrepancy - the fact that you're seeing "2,252,470 bytes (2.3 MB on disk)" in the Apple info box, but "2.1 MB" on the Dataverse side appears to be due to the fact that Apple is interpreting "1 MB" as "1,000,000 bytes"; and Dataverse is (more correctly!) using the "1MB = 1024 * 1024 bytes" definition. What we are doing is more correct. But what Apple is doing is somewhat OK too - both definitions are widely used (unfortunately). And real life users are used to seeing  discrepancies between the 2 definitions by now. 

Can I move it to QA?

kcondon commented:
@landreev One thing I learned from my networking class is the 1,000,000 bytes value is used to count network traffic and the 1024 style is for disk storage. I think. Will need to reference my class notes ;)

kcondon commented:
@pdurbin @landreev So, it looks pretty good -I'm fine with file sizes. However, I did notice one issue:
the old style download all button still presents orig and archival formats when tab file is removed in the current version. While testing shows this exists in 4.20, what is different is when you choose download original, it throws an error on download:
{"status":"ERROR","code":404,"message":"'/api/v1/access/datafile/53792' datafile access error: requested optional service (image scaling, format conversion, etc.) is not supported on this datafile."}

So, kind of a weird issue where part of it preexists but is actually worse now. Maybe the same fix that was applied to the access download all would work here too?

kcondon commented:
@pdurbin @landreev @mheppler I've found a weird, preexisting bug with old download all that is technically out of scope for this issue but I'm wondering whether the same fix for hiding tab download options when tab deleted in current version might be easy to add to it?
The bug is when you have deleted tab file in existing version, have a restricted file with terms of use,  not just terms of access enabled, do not yet have access to restricted file, and choose download all but original version only. Kind of narrow but throws ugly white error screen and nothing is downloaded: {"status":"ERROR","code":404,"message":"'/api/v1/access/datafile/53792' datafile access error: requested optional service (image scaling, format conversion, etc.) is not supported on this datafile."}

landreev commented:
@kcondon What's "old download all"? Do you mean selecting all the checkboxes and clicking "download" in the upper right corner of the files table? 
I'm quite confused by the condition you are describing. If it's something pre-existing, I'm surprised nobody has noticed before.

landreev commented:
@pdurbin @kcondon @mheppler
I'm assuming this is the test dataset in question: https://dataverse-internal.iq.harvard.edu/dataset.xhtml?persistentId=doi%3A10.70122%2FFK2%2FSJSEZX&version=3.1

Yeah, I can see the "original" and "tabular" options under the old style multi-download button. I assumed earlier, that the way it used to work, there was never a real "download all", i.e. it was always a download of the specified list of files from the list of the selected checkboxes... and that way the above condition would never happen, in a version that has no tabular files. But we must have added an optimization at some point - I now have some vague memories of it - for that "select all" checkbox, that actually bypasses the checkboxes and goes to the complete list of files when "all" is selected... So yes, it is probably the same logic error - searching for tabular files in the entire dataset; instead of the current version. And could probably be fixed just as easily. 
Whether it should be fixed as part of this PR, seeing how it is pre-existing and in a different code section - I don't have a strong opinion.

landreev commented:
@pdurbin @kcondon @mheppler

Oh, and here's the reason you get that weird API error: because one of the files is restricted, and there are 2 files total, in the end there is only 1 file in the list - and Dataverse is smart enough to then determine that this is NOT a multiple file download case; so it redirects it to the single file download API. apparently it is not smart enough to then drop the "format=original" parameter - and the single download API treats that as a fatal error, when a format that's not applicable to the current file is requested. 
This is annoying - but is indeed a fairly edge case....

kcondon commented:
@landreev @pdurbin @mheppler Leonid, thanks for confirming! I don't believe it is technically in scope but the question I was asking is whether it would be easy to adopt the solution for not showing download original/archive on versions that do not have tabular files, like Phil did already. If not, I'll just open a new issue.

landreev commented:
@kcondon

>  @pdurbin @mheppler Leonid, thanks for confirming! I don't believe it is technically in scope but the question I was asking is whether it would be easy to adopt the solution for not showing download original/archive on versions that do not have tabular files, like Phil did already. If not, I'll just open a new issue.

I feel like it's worth fixing the "don't show the format choice when there are no tab. files left in the version" part. The fix would be the exact. same approach; we'll just need to find where to apply it in the old code.

landreev commented:
OK, it looks like this situation is NOT even unique to the "select all checkboxes" scenario. In the old setup, it really looks like we ALWAYS present  the "original vs. tabular" choice whenever there are any tab. files in the dataset. Even if the selected checkboxes are not tabular; even if there's only one checkbox, for a non-tabular file, that is selected (we have an open issue - #6972 for that). 

It really looks like we want to replace this with a check on only the files that are actually selected. It's not a huge deal of an issue - but could be confusing. The fact that we've been talking about it for a while now illustrates that point.

landreev commented:
As I just mentioned to @kcondon and @djbrooke , I'm totally fine with merging this PR; and then using #6972 to fix the pre-existing conditions described above. Any objections?

kcondon commented:
Going once, twice...
